core/src/autogluon/core/utils/feature_selection.py

```diff
@@ -0,0 +1,537 @@
+        importance scores are below threshold until removal candidate configuration that has not yet been tried
+        is encountered. Give higher probability to features whose scores are lower than others when sampling.
+        """
+        time_start = time.time()
+        n_remove = max(1, int(prune_ratio * len(importance_df)))
+        above_threshold_rows = importance_df[(importance_df['importance'] > prune_threshold) | (importance_df['importance'].isna())]
+        below_threshold_rows = importance_df[importance_df['importance'] <= prune_threshold].sort_values(by='importance', axis=0, ascending=True)
+        logger.log(15, f"\tNumber of identified features below prune threshold {round(prune_threshold, 4)}: {len(below_threshold_rows)}/{len(importance_df)}")
+        if len(below_threshold_rows) <= n_remove:
+            acceptance_candidates = above_threshold_rows.index.tolist()
+            self.attempted_removals.add(tuple(below_threshold_rows.index))
+            return acceptance_candidates
+
+        # Try removing features with lowest importance first
+        removal_candidate_rows = below_threshold_rows[:n_remove]
+        removal_candidates = tuple(removal_candidate_rows.index)
+        if removal_candidates not in self.attempted_removals:
+            acceptance_candidates = importance_df[~importance_df.index.isin(removal_candidates)].index.tolist()
+            self.attempted_removals.add(removal_candidates)
+            return acceptance_candidates
+
+        sample_weights = [1/i for i in range(1, len(below_threshold_rows)+1)] if weighted else None
+        while time_budget - (time.time() - time_start) > 0:
+            random_state = self.rng.integers(low=0, high=1e5)
+            removal_candidate_rows = below_threshold_rows.sample(n=n_remove, random_state=random_state, replace=False, weights=sample_weights)
+            removal_candidates = tuple(removal_candidate_rows.index)
+            if removal_candidates not in self.attempted_removals:
+                acceptance_candidates = importance_df[~importance_df.index.isin(removal_candidates)].index.tolist()
+                self.attempted_removals.add(removal_candidates)
+                return acceptance_candidates
+        return importance_df.index.tolist()
+
+    def compute_expected_fi_time_single(self, X_fi: pd.DataFrame, model_predict_time: float, n_subsample: int, n_total_sample: int) -> float:
+        """
+        Return the expected time to compute permutation feature importance for a single feature on n_total_sample datapoints.
+        """
+        n_subsample = min(n_subsample, len(X_fi))
+        n_shuffle = min(np.ceil(n_total_sample / n_subsample).astype(int), self.max_n_shuffle)
+        return self.safety_time_multiplier * model_predict_time * (n_subsample / len(X_fi)) * n_shuffle
+
+    def compute_time_budget_fi(self, X_fi: pd.DataFrame, n_subsample: int, **kwargs):
```

Comment:
why use kwargs when you only use two kwarg values and you know what they are and their defaults?

Reply:
I wrote it this way so I can potentially write extensions of feature pruning logic (ex. bandit based) and be able to use select_features / compute_next_candidate